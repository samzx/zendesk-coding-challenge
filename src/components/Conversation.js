import React from 'react';

class Conversation extends React.Component {

    render() {
        return(
            <div className="conversation">
            {
                this.props.loading ?
                <p>Loading comments...</p> :
                this.props.comments &&
                this.props.comments.map((item, index) => 
                    <div key={`conversation-${index}`}>{item.author_id} {item.body}</div>
                )
            }
            </div>
        );
    }
}

export default Conversation;