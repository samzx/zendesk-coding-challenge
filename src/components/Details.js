import React from "react";

class Details extends React.Component {
  shortenTime = timeString => {
    const date = new Date(timeString);
    if (date.toDateString() == "Invalid Date") return "";
    return date.toLocaleString("en-US");
  };

  createTitleRow = titleArr => {
    return (
      <tr>
        {titleArr.map((title, index) => (
          <td key={`${title}-${index}`}>
            <h4>{title}</h4>
          </td>
        ))}
      </tr>
    );
  };

  createDataRow = dataArr => {
    return (
      <tr>
        {dataArr.map((data, index) => <td key={`${data}-${index}`}>{data}</td>)}
      </tr>
    );
  };

  createTags = tags => {
    return tags.map((tag, index) => (
      <span key={`tag-${index}`}>
        {tag}
        {index + 1 != tags.length && ", "}
      </span>
    ));
  };

  display = ticket => {
    const {
      id,
      created_at,
      submitter_id,
      requester_id,
      assignee_id,
      status,
      type,
      priority,
      subject,
      tags,
      description
    } = ticket;
    return (
      ticket && (
        <div>
          <h3 className="details--id">Ticket {id}</h3>
          <p className="details--created">
            Created {this.shortenTime(created_at)}
          </p>
          <table className="details--table">
            <tbody>
              {this.createTitleRow([
                "Submitter ID",
                "Requester ID",
                "Asignee ID"
              ])}
              {this.createDataRow([submitter_id, requester_id, assignee_id])}
              {this.createTitleRow(["Status", "Type", "Priority"])}
              {this.createDataRow([status, type, priority])}
            </tbody>
          </table>
          <hr className="details--split" />
          <div>
            <h3>Subject</h3>
            <p>{subject}</p>
            <h3>Tags</h3>
            <p>{this.createTags(tags)}</p>
            <h3>Description</h3>
            <p> {description}</p>
          </div>
        </div>
      )
    );
  };

  render() {
    return (
      <div className="details">
        {this.props.data &&
          this.props.currentTicket &&
          this.display(
            this.props.data.tickets.find(
              ticket => ticket.id === this.props.currentTicket
            )
          )}
      </div>
    );
  }
}

export default Details;
