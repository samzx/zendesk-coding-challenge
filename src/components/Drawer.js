import React from 'react';

import Pagination from './Pagination';
import List from './List';

import { url, tickets } from '../app';

const initialPage = 1;

class Drawer extends React.Component {

    state = {
        currentPage: 0,
        loading: false,
    }

    fetchListings = (page) => {
        if(this.state.loading || this.state.currentPage === page) return;
        this.setState({ currentPage: page, loading: true });
        fetch(url + tickets + `/${page}`, {
            method: 'GET',
        })
        .then((res) => {
            if(res.ok) {
                return res.json();
            } else {
                throw res;
            }
        })
        .then((json) => {
            this.setState({ loading: false });
            this.props.setData(json);
        })
        .catch((e) => {
            this.setState({ errors: `${e.status} ${e.statusText}`, loading: false });
            console.log(e);
        });
    }

    componentDidMount() {
        this.fetchListings(initialPage);
    }

    render() {
        const { errors, loading } = this.state;
        return(
            <div className="drawer">
            { 
                loading &&
                <div className="loading" >
                    <p>{"Loading..."}</p> 
                </div>
            }
            {
                errors && <p className="loading" >{errors}</p>
            }
                <div className="banner" />
                <List
                    data={this.props.data}
                    currentPage={this.state.currentPage}
                    currentTicket={this.props.currentTicket}
                    setTicket={this.props.setTicket}
                    loading={this.state.loading}
                />
                <Pagination
                    data={this.props.data}
                    currentPage={this.state.currentPage}
                    fetchListings={this.fetchListings}
                />
            </div>
        );
    }
}

export default Drawer;