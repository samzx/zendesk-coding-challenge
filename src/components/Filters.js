import React from 'react';

class Filters extends React.Component {
    render() {
        return(
            <div className="filters" >
                Filters
            </div>
        );
    }
}

export default Filters;