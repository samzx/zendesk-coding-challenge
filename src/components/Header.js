import React from 'react';

class Header extends React.Component {
    render(){
        return(
            <div className="header" >
                <h1>Dashboard</h1>
            </div>
        );
    }
}

export default Header;