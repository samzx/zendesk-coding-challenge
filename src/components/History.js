import React from 'react';

class History extends React.Component {

    render() {
        return(
            <div className="history">
            {
                this.props.comments &&
                this.props.comments.map((item, index) => 
                    <div key={`history-${index}`}>{item.body}</div>
                )
            }
            </div>
        );
    }
}

export default History;