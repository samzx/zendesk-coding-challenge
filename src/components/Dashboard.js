import React from "react";
import ReactDOM from "react-dom";
import Modal from "react-modal";

import Header from "./Header";
import Drawer from "./Drawer";
import Ticket from "./Ticket";

const customStyles = {
  content: {
    maxWidth: "720px",
    margin: "0 auto",
    overflow: "inherit",
    top: "0",
    left: "0",
    right: "0",
    bottom: "0",
    padding: "0"
  }
};

class Dashboard extends React.Component {
  state = {
    showModal: false,
    data: null,
    loading: false,
    errors: null,
    currentTicket: null
  };

  setTicket = id => {
    this.setState({ currentTicket: id, showModal: true });
  };

  setData = data => {
    this.setState({ data });
  };

  render() {
    const { data } = this.state;
    return (
      <div className="dashboard">
        <Drawer
          data={data}
          setData={this.setData}
          currentTicket={this.state.currentTicket}
          setTicket={this.setTicket}
          fetchListings={this.fetchListings}
        />
        <Modal
          ariaHideApp={false}
          id="the-react-modal"
          isOpen={this.state.showModal}
          shouldCloseOnOverlayClick={true}
          style={customStyles}
          onRequestClose={() => {
            this.setState({ showModal: false });
          }}
        >
          <div className="modal-header">
            <span
              className="modal-close"
              onClick={() => this.setState({ showModal: false })}
            >
              {"ğŸ”™"}
            </span>
          </div>
          <div className="banner" />
          <Ticket currentTicket={this.state.currentTicket} data={data} />
        </Modal>
      </div>
    );
  }
}

export default Dashboard;
