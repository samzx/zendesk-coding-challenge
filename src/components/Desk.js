import React from 'react';

import Details from './Details';
import History from './History';

import { url, comments } from '../app';

class Desk extends React.Component {
    state = {
        comments: null,
    }

    componentDidMount() {
        this._mounted = true;
        fetch(url + comments + `/${this.props.currentTicket}`, {
            method: 'GET',
        })
        .then((res) => {
            if(res.ok) {
                return res.json();
            } else {
                throw res;
            }
        })
        .then((json) => {
            if(!this._mounted) return;
            this.setState({ comments: json.comments });
        })
        .catch((e) => {
            if(!this._mounted) return;
            this.setState({ errors: `${e.status} ${e.statusText}`, loading: false });
            console.log(e);
        });
    }

    componentWillUnmount() {
        this._mounted = false;
    }

    render() {
        return(
            <div className="desk">
                <Details
                    currentTicket={this.props.currentTicket}
                    data={this.props.data}
                />
                <History comments={this.state.comments} />
            </div>
        );
    }
}

export default Desk;