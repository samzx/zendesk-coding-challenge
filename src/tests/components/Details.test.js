import React from "react";
import { shallow } from "enzyme";

import Details from "../../components/Details";
import { toReadableTime } from "../../components/Details";
import { response, firstElement } from "../fixtures/response.fixtures";

describe("<Details/>", () => {
  it("should render Details correctly with no data, no ticket selected", () => {
    const wrapper = shallow(<Details />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render Details correctly with data, no ticket selected", () => {
    const wrapper = shallow(<Details data={response} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render Details correctly with data, with ticket 276 selected", () => {
    const wrapper = shallow(
      <Details data={response} currentTicket={firstElement.id} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe("<Details />'s time display", () => {
  it("should shorten time to be easier to read given a valid date string", () => {
    const wrapper = shallow(<Details />);
    expect(wrapper.instance().shortenTime("2018-06-07T13:13:19Z")).toEqual(
      "6/7/2018, 11:13:19 PM"
    );
  });
  it("should not show time given a invalid date string", () => {
    const wrapper = shallow(<Details />);
    expect(wrapper.instance().shortenTime("abcdefg")).toEqual("");
  });
});
