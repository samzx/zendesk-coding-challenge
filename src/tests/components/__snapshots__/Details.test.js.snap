// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Details/> should render Details correctly with data, no ticket selected 1`] = `
<div
  className="details"
/>
`;

exports[`<Details/> should render Details correctly with data, with ticket 276 selected 1`] = `
<div
  className="details"
>
  <div>
    <h3
      className="details--id"
    >
      Ticket 
      276
    </h3>
    <p
      className="details--created"
    >
      Created 
      6/7/2018, 11:13:13 PM
    </p>
    <table
      className="details--table"
    >
      <tbody>
        <tr>
          <td
            key="Submitter ID-0"
          >
            <h4>
              Submitter ID
            </h4>
          </td>
          <td
            key="Requester ID-1"
          >
            <h4>
              Requester ID
            </h4>
          </td>
          <td
            key="Asignee ID-2"
          >
            <h4>
              Asignee ID
            </h4>
          </td>
        </tr>
        <tr>
          <td
            key="365396398731-0"
          >
            365396398731
          </td>
          <td
            key="365396398731-1"
          >
            365396398731
          </td>
          <td
            key="365396398731-2"
          >
            365396398731
          </td>
        </tr>
        <tr>
          <td
            key="Status-0"
          >
            <h4>
              Status
            </h4>
          </td>
          <td
            key="Type-1"
          >
            <h4>
              Type
            </h4>
          </td>
          <td
            key="Priority-2"
          >
            <h4>
              Priority
            </h4>
          </td>
        </tr>
        <tr>
          <td
            key="open-0"
          >
            open
          </td>
          <td
            key="question-1"
          >
            question
          </td>
          <td
            key="urgent-2"
          >
            urgent
          </td>
        </tr>
      </tbody>
    </table>
    <hr
      className="details--split"
    />
    <div>
      <h3>
        Subject
      </h3>
      <p>
        commodo excepteur cupidatat aliquip quis
      </p>
      <h3>
        Tags
      </h3>
      <p>
        <span
          key="tag-0"
        >
          aute
          , 
        </span>
        <span
          key="tag-1"
        >
          incididunt
          , 
        </span>
        <span
          key="tag-2"
        >
          occaecat
        </span>
      </p>
      <h3>
        Description
      </h3>
      <p>
         
        Irure dolore consequat consectetur deserunt id ut ad aliquip ex elit ipsum eu. Mollit cupidatat consequat magna pariatur tempor Lorem. Ut cillum fugiat tempor ad fugiat labore et culpa.

Id in ipsum aliquip nulla est aliquip labore magna occaecat irure. Tempor amet ad eiusmod id. In occaecat dolore labore dolor laboris ullamco sit. Labore enim occaecat sunt aliqua amet duis. Tempor commodo sint ut dolore cupidatat elit ad deserunt minim do veniam est aute. Ex Lorem laborum voluptate aliqua fugiat excepteur do. Sunt non labore incididunt cillum amet cupidatat commodo id consectetur do Lorem reprehenderit ex aliqua.
      </p>
    </div>
  </div>
</div>
`;

exports[`<Details/> should render Details correctly with no data, no ticket selected 1`] = `
<div
  className="details"
/>
`;
